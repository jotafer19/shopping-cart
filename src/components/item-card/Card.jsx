import PropTypes from "prop-types"
import styles from "./Card.module.css"
import { useState } from "react"
import StarRating from "../star-rating/StarRating"
import { useOutletContext } from "react-router-dom"

export default function Card({ item, handleMessageContainer }) {
    const [value, setValue] = useState(1)
    const [cartItems, setCartItems] = useOutletContext()

    function handleAddItem(event, item) {
        event.preventDefault()
        handleMessageContainer()
        if (cartItems.some(cartItem => cartItem.id === item.id)) {
            const newCart = cartItems.map(cartItem => {
                if (cartItem.id === item.id) {
                    return { ...cartItem, count: cartItem.count + value };
                } else {
                    return cartItem;
                }
            })
            setCartItems(newCart);
        } else {
            setCartItems([
                ...cartItems,
                { id: item.id, name: item.title, image: item.image, count: value, price: item.price }
            ])
        }
    }

    function handleLessButton() {
        if (value <= 1) return;
        setValue(prevNumber => prevNumber - 1)
    }

    function handleMoreButton() {
        setValue(prevNumber => prevNumber + 1)
    }

    return (
        <div className={styles.item}>
            <div className={styles["image-container"]}>
                <img src={item.image} alt="Product image" />
            </div>
            <div className={styles.info}>
                <div className={styles.title}>{item.title}</div>
                <div>
                    <StarRating rating={Number(item.rating.rate)}/>
                    <span>({item.rating.count})</span>
                </div>
                <div className={styles.price}>${item.price}</div>
            </div>
            <form className={styles.form} action="#" onSubmit={(event) => handleAddItem(event, item)}>
                <div className={styles.count}>
                    <button className={styles["change-btn"]} onClick={handleLessButton} type="button">-</button>
                    <p>{value}</p>
                    <button className={styles["change-btn"]} onClick={handleMoreButton} type="button">+</button>
                </div>
                <button className={styles["add-btn"]} type="submit">Add to card</button>
            </form>
        </div>
    )
}

Card.propTypes = {
    item: PropTypes.object,
    handleMessageContainer: PropTypes.func,
}