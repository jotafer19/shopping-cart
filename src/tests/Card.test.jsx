import { expect, it, vi, describe } from "vitest";
import { render, screen } from "@testing-library/react";
import Card from "../components/item-card/Card";
import { createContext, useContext } from "react";
import userEvent from "@testing-library/user-event";

vi.mock('react-router-dom', () => ({
    ...vi.importActual('react-router-dom'),
    useOutletContext: () => useContext(MockOutletContext),
  }));

const MockOutletContext = createContext();

function MockOutletContextProvider({ children, value }) {
    return (
      <MockOutletContext.Provider value={value}>
        {children}
      </MockOutletContext.Provider>
    );
  }

const item = { id: 1, name: "foo", image: "#", rating: { rate: 3.1 }, count: 1, price: 10 };

const onClick = vi.fn();

describe("test Card component", () => {
  it("test render", () => {
    const { container } = render(
      <MockOutletContextProvider value={[[], () => {}]}>
        <Card item={item} handleMessageContainer={onClick} />
      </MockOutletContextProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should render buttons to change item count and add to cart", () => {
    render (
        <MockOutletContextProvider value={[[], () => {}]}>
            <Card item={item} handleMessageContainer={onClick} />
        </MockOutletContextProvider>
    )

    expect(screen.getByRole("button", { name: "+" })).toBeInTheDocument()
    expect(screen.getByRole("button", { name: "-" })).toBeInTheDocument()
    expect(screen.getByRole("button", { name: "Add to cart" })).toBeInTheDocument()
  })

  it("more item button should increment item count", async () => {
    const user = userEvent.setup()

    render (
        <MockOutletContextProvider value={[[], () => {}]}>
            <Card item={item} handleMessageContainer={onClick} />
        </MockOutletContextProvider>
    )

    const button = screen.getByRole("button", { name: "+" });
    const countDisplay = screen.getByTestId("item-count")
    expect(countDisplay.textContent).toBe("1")

    await user.click(button)
    
    expect(countDisplay.textContent).toBe("2")
  })

  it("less item button should decrease item count", async () => {
    const user = userEvent.setup()

    render (
      <MockOutletContextProvider value={[[], () => {}]}>
        <Card item={item} handleMessageContainer={onClick} />
      </MockOutletContextProvider> 
    )

    const moreButton = screen.getByRole("button", { name: "+" });
    const lessButton = screen.getByRole("button", { name: "-" });
    const countDisplay = screen.getByTestId("item-count")

    await user.click(moreButton)
    expect(countDisplay.textContent).toBe("2")

    await user.click(lessButton)

    expect(countDisplay.textContent).toBe("1")
  })

  it("should call the onClick function when clicked", async () => {
    const user = userEvent.setup()

    render (
      <MockOutletContextProvider value={[[], () => {}]}>
        <Card item={item} handleMessageContainer={onClick} />
      </MockOutletContextProvider> 
    )

    const button = screen.getByRole("button", { name: "Add to cart" });

    await user.click(button);

    expect(onClick).toHaveBeenCalled()
  })
});
