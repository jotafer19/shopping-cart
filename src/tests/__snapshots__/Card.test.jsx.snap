// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test Card component > test render 1`] = `
<div>
  <div
    class="_item_b07fc0"
  >
    <div
      class="_image-container_b07fc0"
    >
      <img
        alt="Product image"
        src="#"
      />
    </div>
    <div
      class="_info_b07fc0"
    >
      <div
        class="_title_b07fc0"
      />
      <div>
        <span
          class="style-module_starRatingWrap__q-lJC"
          style="direction: ltr;"
        >
          <span
            aria-hidden="true"
            class="style-module_simpleStarRating__nWUxf react-simple-star-rating"
          >
            <span
              class="style-module_emptyIcons__Bg-FZ empty-icons"
              style="color: rgb(204, 204, 204);"
            >
              <svg
                class="star-svg"
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="star-svg"
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="star-svg"
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="star-svg"
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="star-svg"
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
            <span
              class="style-module_fillIcons__6---A filled-icons"
              style="left: 0px; color: rgb(255, 188, 11); width: 62%;"
              title="3.1 out of 5"
            >
              <svg
                class="star-svg"
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="star-svg"
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="star-svg"
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="star-svg"
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <svg
                class="star-svg"
                fill="currentColor"
                height="20"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
            </span>
          </span>
        </span>
        <span>
          (
          )
        </span>
      </div>
      <div
        class="_price_b07fc0"
      >
        $
        10
      </div>
    </div>
    <form
      action="#"
      class="_form_b07fc0"
    >
      <div
        class="_count_b07fc0"
      >
        <button
          class="_change-btn_b07fc0"
          type="button"
        >
          -
        </button>
        <p
          data-testid="item-count"
        >
          1
        </p>
        <button
          class="_change-btn_b07fc0"
          type="button"
        >
          +
        </button>
      </div>
      <button
        class="_add-btn_b07fc0"
        type="submit"
      >
        Add to cart
      </button>
    </form>
  </div>
</div>
`;
